### ISSUE TITLE
Driver stopped working in 0.32.10

### ISSUE THREAD

**Issue by Nikos410** [metabase/metabase#10595]
**Describe the bug**
@pupupulp found out that our Firebird driver for Metabase [does not work with the latest release](https://github.com/evosec/metabase-firebird-driver/issues/1), 0.32.10. The driver is detected just fine, but when establishing a connection, it can not find the required JDBC driver, which is included in the driver's jar file.

**To Reproduce**
1. Install Metabase 0.32.10
2. Install the [Firebird driver](https://github.com/evosec/metabase-firebird-driver) by downloading the latest release and dropping the .jar in the plugins directory.
3. Try establishing a connection to a Firebird database (Just enter fake values, you don't need to set up a real database to reproduce this)
4. See error `No suitable driver found for jdbc:firebirdsql://HOST:PORT/DB`

5. Try the same with metabase 0.32.9 and it works

**Expected behavior**
Metabase should find Jaybird, the Firebird JDBC driver, which is included in the driver jar file.

**Information about your Metabase Installation:**
- Your browser and the version: Firefox 68.0.1

- Your operating system: Ubuntu

- Your databases: Firebird

- Metabase version: 0.32.10

- Metabase hosting environment: Linux/OpenSuse with Java 1.8.0_201

- Metabase internal database: H2 (default)

**Severity**
Blocking an update to version 0.32.10

**Comment by Nikos410** [comment]
I just checked, the same issue exists with v0.33.0-RC1.

**Comment by camsaul** [comment]
@Nikos410 

When Metabase first launches it logs whether or not it can load each of the various drviers, and, if not, the reasons, why. Can you post the logs from when Metabase first launches?

Specifically I'm looking for lines that look like these (with

```
INFO plugins.lazy-loaded-driver :: Registering lazy loading driver :firebird...
```

That's what the log messages should look like if things work as expected. Since things aren't working as expected, the log messages should explain what the problem was.

**Comment by Nikos410** [comment]
@camsaul 

During startup this is the only log output regarding the firebird driver:
```
08-16 08:40:38 INFO plugins.lazy-loaded-driver :: Registering lazy loading driver :firebird...
08-16 08:40:38 INFO metabase.driver :: Registered driver :firebird (parents: [:sql-jdbc]) 🚚
```
I've uploaded the full log to [pastebin](https://pastebin.com/3ecBsHMu)

The driver itself is registered fine, I can select FirebirdSQL as a database type. When I click on next after entering the connection properties I get this log output:

```
08-16 08:48:49 INFO metabase.driver :: Initializing driver :firebird...
08-16 08:48:49 INFO plugins.classloader :: Added URL file:/METABASE_DIR/plugins/firebird.metabase-driver.jar to classpath
08-16 08:48:49 DEBUG plugins.init-steps :: Loading plugin namespace metabase.driver.firebird...
08-16 08:48:49 INFO metabase.driver :: Registered driver :firebird (parents: [:sql-jdbc]) 🚚
08-16 08:48:49 DEBUG middleware.log :: POST /api/setup/validate 400 33 ms (0 DB calls) 
{:errors {:dbname nil}}

08-16 08:48:49 DEBUG middleware.log :: POST /api/setup/validate 400 10 ms (0 DB calls) 
{:errors {:dbname "No suitable driver found for jdbc:firebirdsql://localhost:3050/testing"}}
```

**Comment by Nikos410** [comment]
Hey @camsaul !

Are there any news on what might be causing this?

**Comment by camsaul** [comment]
Sorry @Nikos410, I'm not really sure. I honestly had not idea there was a 3rd-party Firebird driver for Metabase before now. I'll add some better debug logging to the next release to get to the bottom of this issue

**Comment by Nikos410** [comment]
Hey @camsaul , sorry for taking so long to try this out. Using the jar for 0.33.2 I get this output when I click on "next" after entering the firebird connection parameters:

```
09-11 14:57:49 DEBUG middleware.log :: POST /api/util/password_check 200 1.1 ms (0 DB calls) Jetty threads: 4/50 (3 idle, 0 queued) (41 total active threads) Queries in flight: 0
09-11 14:58:36 INFO metabase.driver :: Initializing driver :firebird...
09-11 14:58:36 INFO plugins.classloader :: Added URL file:/metabase_path/plugins/firebird.metabase-driver.jar to classpath
09-11 14:58:36 DEBUG plugins.init-steps :: Loading plugin namespace metabase.driver.firebird...
09-11 14:58:36 INFO metabase.driver :: Registered driver :firebird (parents: [:sql-jdbc]) 🚚
09-11 14:58:36 ERROR driver.util :: Database connection error
java.lang.ExceptionInInitializerError
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at clojure.lang.RT.classForName(RT.java:2211)
	at clojure.lang.RT.classForName(RT.java:2220)
	at clojure.lang.RT.loadClassForName(RT.java:2239)
	at clojure.lang.RT.load(RT.java:449)
	at clojure.lang.RT.load(RT.java:424)
	at clojure.core$load$fn__6839.invoke(core.clj:6126)
	at clojure.core$load.invokeStatic(core.clj:6125)
	at clojure.core$load.doInvoke(core.clj:6109)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at clojure.core$load_one.invokeStatic(core.clj:5908)
	at clojure.core$load_one.invoke(core.clj:5903)
	at clojure.core$load_lib$fn__6780.invoke(core.clj:5948)
	at clojure.core$load_lib.invokeStatic(core.clj:5947)
	at clojure.core$load_lib.doInvoke(core.clj:5928)
	at clojure.lang.RestFn.applyTo(RestFn.java:142)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$load_libs.invokeStatic(core.clj:5985)
	at clojure.core$load_libs.doInvoke(core.clj:5969)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:667)
	at clojure.core$require.invokeStatic(core.clj:6007)
	at clojure.core$require.doInvoke(core.clj:6007)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at metabase.plugins.classloader$require.invokeStatic(classloader.clj:99)
	at metabase.plugins.classloader$require.doInvoke(classloader.clj:89)
	at clojure.lang.RestFn.invoke(RestFn.java:408)
	at metabase.plugins.init_steps$fn__65964.invokeStatic(init_steps.clj:22)
	at metabase.plugins.init_steps$fn__65964.invoke(init_steps.clj:20)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invokeStatic(init_steps.clj:32)
	at metabase.plugins.init_steps$do_init_steps_BANG_.invoke(init_steps.clj:27)
	at metabase.plugins.lazy_loaded_driver$make_initialize_BANG_$fn__66042.invoke(lazy_loaded_driver.clj:59)
	at clojure.lang.MultiFn.invoke(MultiFn.java:229)
	at metabase.driver$initialize_if_needed_BANG_.invokeStatic(driver.clj:293)
	at metabase.driver$initialize_if_needed_BANG_.invoke(driver.clj:277)
	at metabase.driver$the_initialized_driver.invokeStatic(driver.clj:300)
	at metabase.driver$the_initialized_driver.invoke(driver.clj:296)
	at metabase.driver$dispatch_on_initialized_driver.invokeStatic(driver.clj:308)
	at metabase.driver$dispatch_on_initialized_driver.doInvoke(driver.clj:303)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__19324.invoke(util.clj:30)
	at metabase.util$do_with_timeout$fn__6283.invoke(util.clj:334)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: clojure.lang.Var$Unbound cannot be cast to clojure.lang.MultiFn
	at metabase.driver.firebird__init.load(Unknown Source)
	at metabase.driver.firebird__init.<clinit>(Unknown Source)
	... 53 more
09-11 14:58:36 DEBUG middleware.log :: POST /api/setup/validate 400 39.2 ms (0 DB calls) 
{:errors {:dbname ""}}

09-11 14:58:36 ERROR driver.util :: Database connection error
java.sql.SQLException: No suitable driver found for jdbc:firebirdsql://localhost:3050/firebird-testing
	at java.sql.DriverManager.getConnection(DriverManager.java:689)
	at java.sql.DriverManager.getConnection(DriverManager.java:208)
	at clojure.java.jdbc$get_driver_connection.invokeStatic(jdbc.clj:271)
	at clojure.java.jdbc$get_driver_connection.invoke(jdbc.clj:250)
	at clojure.java.jdbc$get_connection.invokeStatic(jdbc.clj:411)
	at clojure.java.jdbc$get_connection.invoke(jdbc.clj:274)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1093)
	at clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1075)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1164)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1126)
	at clojure.java.jdbc$query.invokeStatic(jdbc.clj:1142)
	at clojure.java.jdbc$query.invoke(jdbc.clj:1126)
	at metabase.driver.firebird$fn__203.invokeStatic(firebird.clj:45)
	at metabase.driver.firebird$fn__203.invoke(firebird.clj:43)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at metabase.driver.util$can_connect_with_details_QMARK_$fn__19324.invoke(util.clj:30)
	at metabase.util$do_with_timeout$fn__6283.invoke(util.clj:334)
	at clojure.core$binding_conveyor_fn$fn__5754.invoke(core.clj:2030)
	at clojure.lang.AFn.call(AFn.java:18)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
09-11 14:58:36 DEBUG middleware.log :: POST /api/setup/validate 400 11.8 ms (0 DB calls) 
{:errors {:dbname "No suitable driver found for jdbc:firebirdsql://localhost:3050/firebird-testing"}}
```

**Comment by rylonian** [comment]
I am also having this same issue when upgrading from 0.32.7 to 0.33.0 or 0.33.2.
However, we use SQL Server in our environment instead of Firebird.  I can post logs later because I will need to reproduce it in a test environment.

**Comment by kumajaya** [comment]
Firebird driver update needed related to https://github.com/metabase/metabase/commit/e7100252205e01c16daa431f52539dbf8d26c438#diff-d77ef6d7d60b04c2447263a86bfaeabf

**Comment by kumajaya** [comment]
Yes, sure. Metabase lazy load Firebird driver: https://github.com/metabase/metabase/wiki/Writing-a-Driver:-Packaging-a-Driver-&-Metabase-Plugin-Basics#lazy-loading

**Comment by camsaul** [comment]
Closing as resolved since your Firebird Metabase driver needs to be updated and this is not a bug with Metabase itself.

**Comment by Nikos410** [comment]
Oh, I already fixed our driver, totally forgot to close this issue - sorry.

But I do have one suggestion regarding this issue: Since the error message I got did not really say anything about my driver missing a required function, and there was no info about Metabase making breaking changes to the driver interface (at least none that I saw), it was really hard for me to find out what was causing the issue. If @kumajaya had not told me that I had to implement date-add, I would still be searching.

Maybe Metabase could verify drivers when loading them and log a warning if there are any missing functions instead of saying that the driver was loaded successfully. Is that possible @camsaul ?
